sympy/core/equation.py

```diff
@@ -0,0 +1,359 @@
+'''
+from sympy.simplify import simplify
+from sympy.core import Function
+from sympy.functions import *
+from sympy import functions
+from sympy import factorial
+'''
+
+#from sympy import * # Replace with more selective import when folded into SymPy
+
+class Equation(Basic):
+    """
+    This class defines an equation with a left-hand-side (lhs) and a right-hand-side (rhs) connected by an operator (e.g. $p*V = n*R*T$).
+    
+    Explanation
+    ===========
+    This class defines relations that all high school and college students would recognize as mathematical equations. 
+    At present only the "=" relation operator is recognized.
+    
+    This class is intended to allow using the mathematical tools in SymPy to rearrange equations and perform algebra
+    in a stepwise fashion. In this way more people can successfully perform algebraic rearrangements without stumbling
+    over missed details such as a negative sign.
+    
+    __Note__ that this module imports Sympy into its namespace so there is no need to import Sympy separately.
+    
+    Create an equation with the call ``Equation(lhs,rhs,relation_operator)``, where ``lhs`` and ``rhs`` are any valid Sympy
+    expression. ``relation_operator`` defaults to the string "=" if not supplied. Currently,"=" is the only valid option.
+    ``Eqn(...)`` is a synonym for ``Equation(...)``.
+    
+    Examples
+    ========
+    >>> var('a b c')
+    >>> equ(a,b/c)
+    a=b/c
+    >>> t=equ(a,b/c)
+    >>> t
+    a=b/c
+    >>> t*c
+    a*c=b
+    >>> c*t
+    a*c=b
+    >>> exp(t)
+    exp(a)=exp(b/c)
+    >>> exp(log(t))
+    a=b/c
+
+    Integration can only be performed on one side at a time.
+    >>> q=Eqn(a*c,b/c)
+    >>> integrate(q,b,side='rhs')
+    b**2/(2*c)
+    >>> integrate(q,b,side='lhs')
+    a*b*c
+
+    >>> # Make a pretty statement of integration from an equation 
+    >>> Eqn(Integral(q.lhs,b),integrate(q,b,side='rhs'))                        
+    Integral(a*c, b)=b**2/(2*c)
+    >>> # This is duplicated by the convenience function self.integ 
+    >>> q.integ(b)                                                              
+    Integral(a*c, b)=b**2/(2*c)
+
+    SymPy's solvers do not understand these equations. They expect an expression that the solver assumes = 0.
+    Thus to use the solver the equation must be rearranged so that all non-zero symbols are on one side. Then
+    just the non-zero symbolic side is passed to ``solve()``.
+    >>> t2 = t-t.rhs
+    >>> t2
+    a-b/c=0
+    >>> solve(t2.lhs,c)
+    [b/a]
+    """
+
+    def __new__(cls, lhs, rhs, relop='='):
+        if not(relop == '='):
+           raise NotImplementedError('"=" is the only relational operator presently supported in Equations.')
+        lhs = _sympify(lhs)
+        rhs = _sympify(rhs)
+        if (lhs.is_number) and (rhs.is_number) and not (lhs == rhs):
+           print('WARNING: did your really mean to define unequal numbers as equal? ' + str(lhs) +'='+ str(rhs))
```

Comment:
We don't want this warning. In sympy `lhs==rhs` would (mostly) only be true if both sides were the exact same expression. There wouldn't be much point in having two sides of the equation if they are the exact same expression:
```
In [5]: Equation(exp(I*pi/4), (1+I)/sqrt(2))                                                                                      
WARNING: did your really mean to define unequal numbers as equal? exp(I*pi/4)=sqrt(2)*(1 + I)/2
Out[5]: exp(I*pi/4)=sqrt(2)*(1 + I)/2
```
Or perhaps
```
In [10]: Equation(pi*(I+2), pi*I+2*pi)                                                                                            
WARNING: did your really mean to define unequal numbers as equal? pi*(2 + I)=2*pi + I*pi
Out[10]: pi*(2 + I)=2*pi + I*pi
```

Reply:
A better check would be `lhs.equals(rhs)`. But I think we should either disallow this completely (exception), or leave it alone. I'm not a fan of warnings (also warnings should use `warnings.warn`, not `print`).

Reply:
Using `.equals` can be very expensive (it *starts* by calling `simplify`).

Reply:
I understand the problem with the current implementation. However, after watching myself and lots of other relatively computer and math savvy scientists code and use tools like this, I am convinced that some kind of sanity  check will save many people a lot of effort. I think the issue is sloppy typing more than anything else.

Too avoid the overhead of `simplify`, I could make it so that calls within the class skip this test, by adding a keyword `check=yes` or `check=no`. Internal calls would all use `check=no`. I think by doing this, I can limit the checking only to manually input equations. It would also give users the option of overriding the check.

I was unaware that SymPy implemented `warnings.warn`. I can switch to that.

Reply:
Am I correct in understanding that `warnings` is imported only when needed? What I think of as lazy imports, but I think that is the wrong term.

Reply:
There are a few modules that already import warnings.

I don't think that there should be a warning here though. You can check using `.as_Boolean().simplify()` and an alias could be made to do that and return True, False, None. Also it is useful sometimes to be able to create invalid objects e.g. for display purposes.

Reply:
If I understand this last suggestion, it requires the user to do something for sanity checking to occur. My opinion is that from an end user perspective, it would be better to have the sanity checking happen when a user creates an expression by default. 

I agree that some way to turn checking off, so that "invalid" expressions can be created is important. However, I am not yet convinced that we should skip providing a simple sanity check. What problem or problems will such a check cause as long as it only happens on user input?

Reply:
In my opinion, why not drop every check and allow `0 = 1`?
`0 = 1` may not make sense for some of you people, but for equation manipulation, I think that it can be well defined like an operation that translates `0 = 1` to `1 = 2`, `a = a+1`.

If someone wants to simulate how some equation arrives to such contradiction, it would also be a good option to allow `0 = 1`, and hence reduce the computational overhead for figuring it out.

Reply:
Without risking a significant slowdown for some inputs it will only be possible to do this sanity checking in situations that are very limited that will probably not help anyone find an actual bug.

The vast majority of the time this class will be used with a symbol in it somewhere and this check won't even do anything. In the cases where someone does want to create an Equation with both sides numbers then there's a good chance that they are doing it for display or something else and will just be annoyed by this.

I think it's better to make it clear in the documentation that it's the user's responsibility to check what they put in but that there are methods for helping with that.

Reply:
@oscarbenjamin 
I agree that most of the time people will be intending to put in expressions with symbols and then nothing would be checked. However, what I have found watching people work with IPython/Jupyter notebooks that they leave a lot of stray definitions around, that trip them up. They also forget about the reserved symbols within SymPy (I, pi ...). A check for the lhs and rhs collapsing to numbers that are not equal will catch some of those types of errors in a very inobtrusive way. Most should never see the warning the way I am envisioning it. 

Compared to most people, I am an obsessive reader of instructions and documentation, but have not read the detailed docs of most of the operations I use in SymPy, SageMath, and Maple. I expect them to behave according to the standards implied by the notation. I only look carefully, when I get unexpected results. Thus I think people are very unlikely to be aware of options for checking that they are putting reasonable expressions in unless they notice that their results are unreasonable, cannot figure out why and then start reading the documentation. In my experience "noticing that results are unreasonable" is something very few people appear to be good at.

@sylee957 
`a = a +1` would not be checked if `a` is a symbol.  So `Eqn(a,a+1)` would not and does not raise a warning. Then the user could try to isolate `a` by subtracting it from both sides and reach the conclusion `0 = 1`. It would still be up to the user to decide if this is what they want.

Were you thinking of `a=a+1`, as the convenient shorthand in high level languages for the two step process below?
```
temp = a
a = temp + 1
```
I was not thinking that the `Equation` type would also include operations like this, only what would generally be regarded as an equation in the form usually used in "on-paper" math.

Reply:
I also don't think that equation is same as assignment, but I just had multiple questions to make sure the equation manipulation looks well defined in some sense, and it is not just 'for kids'.

Reply:
My commit https://github.com/sympy/sympy/pull/19479/commits/14e0a1cab0c811b83706fb591e8e45092fab5ae6 includes a version of the warning that behaves more or less the way I intended it (the output from `warnings.warn` and other options within `warnings` is not very nice). I do not think it will give warnings when it should not and I really do think it will be helpful to end users.